### PR TITLE
[SHELL32] RecycleBin5Enum: Make it C++

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Modernize code.
JIRA issue: [CORE-19595](https://jira.reactos.org/browse/CORE-19595)

## Proposed changes

- Rewrite `RecycleBin5Enum` in C++.

### CODE REVIEW

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+#include <shlwapi.h>
+
+class RecycleBin5File : public IRecycleBinFile
+{
+public:
+    RecycleBin5File(
+        IN IRecycleBin5 *prb,
+        IN LPCWSTR Folder,
+        IN PDELETED_FILE_RECORD pDeletedFile,
+        IN OUT LPWSTR pszFullName);
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:19]: You could take the opportunity to use the new SAL2 annotations (and in the other recyclebin files you've converted)

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+    DWORD dwAttributes = ::GetFileAttributesW(m_FullName);
+    if (dwAttributes == INVALID_FILE_ATTRIBUTES)
+        return HRESULT_FROM_WIN32(GetLastError());
+
+    HANDLE hFile;
+    if (dwAttributes & FILE_ATTRIBUTE_DIRECTORY)
+        hFile = CreateFileW(m_FullName, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    else
+        hFile = CreateFileW(m_FullName, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:106]: Something along these lines, maybe?
```suggestion
    hFile = CreateFileW(m_FullName,
                        GENERIC_READ,
                        FILE_SHARE_READ |
                            ((dwAttributes & FILE_ATTRIBUTE_DIRECTORY) ? (FILE_SHARE_WRITE | FILE_SHARE_DELETE) : 0),
                        NULL,
                        OPEN_EXISTING,
                        (dwAttributes & FILE_ATTRIBUTE_DIRECTORY) ? FILE_FLAG_BACKUP_SEMANTICS : 0,
                        NULL);
```

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+    IN IRecycleBin5 *prb,
+    IN LPCWSTR Folder,
+    IN PDELETED_FILE_RECORD pDeletedFile,
+    OUT IRecycleBinFile **ppFile)
+{
+    if (!ppFile)
+        return E_POINTER;
+
+    LPCWSTR Extension = wcsrchr(pDeletedFile->FileNameW, '.');
+    WCHAR szFullName[MAX_PATH];
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:261]: I'm not fond of this new local buffer, while the original code was able to avoid using one.
Original code was calculating the necessary size of the string, then allocated a buffer of proper size before doing the wsprintf call.
(Also consider using a stringsafe function here?)

In addition, the old code was doing some extra checks on the retrieved "Extension" pointer. Wouldn't they be needed there as well? Or, what about using `PathFindExtensionW()` instead, if that latter function also does some extra "validation" on where the extension must be found?

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+    return S_OK;
+}
+
+class RecycleBin5Enum : public IRecycleBinEnumList
+{
+public:
+    RecycleBin5Enum(
+        IN IRecycleBin5 *prb,
+        IN HANDLE hInfo,
+        IN OUT INFO2_HEADER *pInfo,
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:285]: All the `INFO2_HEADER *` --> `PINFO2_HEADER`

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+    ULARGE_INTEGER FileSize;
+    FileSize.u.LowPart = GetFileSize(m_hInfo, &FileSize.u.HighPart);
+    if (FileSize.u.LowPart == 0)
+        return HRESULT_FROM_WIN32(GetLastError());
+
+    DWORD dwEntries =
+        (DWORD)((FileSize.QuadPart - sizeof(INFO2_HEADER)) / sizeof(DELETED_FILE_RECORD));
+
+    DWORD iEntry = m_dwCurrent, fetched = 0;
+    DELETED_FILE_RECORD *pDeletedFile = (DELETED_FILE_RECORD *)(m_pInfo + 1) + iEntry;
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:373]: All the `DELETED_FILE_RECORD *` --> `PDELETED_FILE_RECORD`

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+    TRACE("(%p, %u, %p, %p)\n", this, celt, rgelt, pceltFetched);
+
+    if (!rgelt)
+        return E_POINTER;
+    if (!pceltFetched && celt > 1)
+        return E_INVALIDARG;
+
+    ULARGE_INTEGER FileSize;
+    FileSize.u.LowPart = GetFileSize(m_hInfo, &FileSize.u.HighPart);
+    if (FileSize.u.LowPart == 0)
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:366]: There's a bug here, because according to MSDN the correct check to see whether there has been an error in the GetFileSize call would have to read something like:
```c
        if ((FileSize.u.LowPart  == INVALID_FILE_SIZE) && (GetLastError() != ERROR_SUCCESS))
```
Otherwise I think you could use GetFileSizeEx :thinking:

This change may be done in a separate PR, or at the very least in a separate commit (since the current code is what was originally present)

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+{
+    if (!ppUnknown)
+        return E_POINTER;
+
+    LPWSTR pszPrefix;
+    HRESULT hr = SHStrDup(szPrefix, &pszPrefix);
+    if (FAILED(hr))
+        return hr;
+
+    INFO2_HEADER *pInfo = (INFO2_HEADER *)MapViewOfFile(hInfoMapped, FILE_MAP_READ, 0, 0, 0);
```
Comment (HBelusca) [inline dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp:437]: I think this whole pre-init code could be put in the RecycleBin5Enum C++ constructor.

Because as it currently is, it's RecycleBin5Enum_Constructor(), not belonging to the class, that calls the MapView..., but it's the `~RecycleBin5Enum()` class destructor that calls the UnMap... routine (+ the other CloseHandles calls).

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp]: Fixed.

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp]: Fixed.

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp]: Fixed by using ATL `CStringW` and `PathFindExtensionW`.

File: dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp]: Used `Init()` helper method.
